mp4parse/src/lib.rs

```diff
@@ -388,6 +392,529 @@ pub enum SampleEntry {
+                    // Total chunk number in 'stsc' could be different to 'stco',
+                    // there could be more chunks at the last 'stsc' record.
+                    match next.first_chunk.checked_add(self.remain_chunk_count) {
+                        Some(r) => (next.first_chunk - 1)..r - 1,
+                        _ => 0..0,
+                    }
+                }
+                _ => 0..0,
+            };
+        }
+    }
+}
+
+/// Calculate numerator * scale / denominator, if possible.
+///
+/// Applying the associativity of integer arithmetic, we divide first
+/// and add the remainder after multiplying each term separately
+/// to preserve precision while leaving more headroom. That is,
+/// (n * s) / d is split into floor(n / d) * s + (n % d) * s / d.
+///
+/// Return None on overflow or if the denominator is zero.
+pub fn rational_scale<T, S>(numerator: T, denominator: T, scale2: S) -> Option<T>
+where
+    T: PrimInt + Zero,
+    S: PrimInt,
+{
+    if denominator.is_zero() {
+        return None;
+    }
+
+    let integer = numerator / denominator;
+    let remainder = numerator % denominator;
+    num_traits::cast(scale2).and_then(|s| match integer.checked_mul(&s) {
+        Some(integer) => remainder
+            .checked_mul(&s)
+            .and_then(|remainder| (remainder / denominator).checked_add(&integer)),
+        None => None,
+    })
+}
+
+pub fn media_time_to_us(time: MediaScaledTime, scale: MediaTimeScale) -> Option<u64> {
+    let microseconds_per_second = 1_000_000;
+    rational_scale::<u64, u64>(time.0, scale.0, microseconds_per_second)
+}
+
+pub fn track_time_to_us<T>(time: TrackScaledTime<T>, scale: TrackTimeScale<T>) -> Option<T>
+where
+    T: PrimInt + Zero,
+{
+    assert_eq!(time.1, scale.1);
+    let microseconds_per_second = 1_000_000;
+    rational_scale::<T, u64>(time.0, scale.0, microseconds_per_second)
+}
+
+#[test]
+fn rational_scale_overflow() {
+    assert_eq!(rational_scale::<u64, u64>(17, 3, 1000), Some(5666));
+    let large = 0x4000_0000_0000_0000;
+    assert_eq!(rational_scale::<u64, u64>(large, 2, 2), Some(large));
+    assert_eq!(rational_scale::<u64, u64>(large, 4, 4), Some(large));
+    assert_eq!(rational_scale::<u64, u64>(large, 2, 8), None);
+    assert_eq!(rational_scale::<u64, u64>(large, 8, 4), Some(large / 2));
+    assert_eq!(rational_scale::<u64, u64>(large + 1, 4, 4), Some(large + 1));
+    assert_eq!(rational_scale::<u64, u64>(large, 40, 1000), None);
+}
+
+#[test]
+fn media_time_overflow() {
+    let scale = MediaTimeScale(90000);
+    let duration = MediaScaledTime(9_007_199_254_710_000);
+    assert_eq!(
+        media_time_to_us(duration, scale),
+        Some(100_079_991_719_000_000)
+    );
+}
+
+#[test]
+fn track_time_overflow() {
+    let scale = TrackTimeScale(44100u64, 0);
+    let duration = TrackScaledTime(4_413_527_634_807_900u64, 0);
+    assert_eq!(
+        track_time_to_us(duration, scale),
+        Some(100_079_991_719_000_000)
+    );
+}
```

Comment:
This is a move of https://github.com/mozilla/mp4parse-rust/blob/1ab30201ad06389d0d0ebe3036418d624e7219f4/mp4parse_capi/src/lib.rs#L2088-L2118 without any modification. I appreciate moving the tests closer to the code they cover.